ppa/archive/management/commands/gale_import.py

```diff
@@ -0,0 +1,207 @@
+        # parser.add_argument(
+        #     '-u', '--update', action='store_true',
+        #     help='Update local content even if source record has not changed.')
+        # parser.add_argument(
+        #     '--progress', action='store_true',
+        #     help='Display a progress bar to track the status of the import.')
+
+    def handle(self, *args, **kwargs):
+
+        if not (kwargs["ids"] or kwargs["csv"]):
+            raise CommandError("A list of IDs or CSV file for is required for import")
+
+        # disconnect signal handler for on-demand indexing, for efficiency
+        # (index in bulk after an update, not one at a time)
+        IndexableSignalHandler.disconnect()
+
+        self.gale_api = GaleAPI()
+        self.stats = Counter()
+        self.script_user = User.objects.get(username=settings.SCRIPT_USERNAME)
+        self.digwork_contentype = ContentType.objects.get_for_model(DigitizedWork)
+
+        # if ids are specified on the command line, create a list
+        # of dictionaries so import will look similar to csv
+        if kwargs["ids"]:
+            to_import = [{"ID": gale_id} for gale_id in kwargs["ids"]]
+        # when csv is specified, load rows into a list of dics
+        elif kwargs["csv"]:
+            to_import = self.load_csv(kwargs["csv"])
+            # load collections when importing from CSV
+            self.load_collections()
+
+        # total is needed for progessbar (if we add it)
+        self.stats["total"] = len(to_import)
+
+        for item in to_import:
+            if self.verbosity >= self.v_normal:
+                # include title in output if present, but truncate since many are long
+                self.stdout.write(
+                    " ".join([item["ID"], truncatechars(item.get("Title", ""), 55)])
+                )
+            self.import_digitizedwork(item["ID"], **item)
+
+        summary = (
+            "\nProcessed {:,d} item{} for import."
+            + "\nImported {:,d}; skipped {:,d}; {:,d} error{}; imported {:,d} page{}."
+        )
+        summary = summary.format(
+            self.stats["total"],
+            pluralize(self.stats["total"]),
+            self.stats["imported"],
+            self.stats["skipped"],
+            self.stats["error"],
+            pluralize(self.stats["error"]),
+            self.stats["pages"],
+            pluralize(self.stats["pages"]),
+        )
+        self.stdout.write(summary)
+
+    def load_collections(self):
+        # load collections from the database and create
+        # a lookup based on the codes used in the spreadsheet
+        self.collections = {}
+        collections = {c.name: c for c in Collection.objects.all()}
+        for code, name in self.collection_codes.items():
+            self.collections[code] = collections[name]
+
+    def load_csv(self, path):
+        """Load a CSV file with items to be imported."""
+        try:
+            with open(path, encoding="utf-8-sig") as csvfile:
```

Comment:
does `utf-8-sig` refer to utf8 with BOM?

Reply:
Yeah — I was getting an error on the first fieldname because it included `\ufeff` ; specifying the encoding handles it properly (it doesn't include the BOM in the content), but I'm not sure if we'll always have this encoding.  I was working with a csv exported by Excel, since the team has been working in Excel spreadsheets — so it might be ok, but not sure.